sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/ProactiveOpenConnectionsProcessor.java

```diff
@@ -0,0 +1,253 @@
+
+public final class ProactiveOpenConnectionsProcessor implements Closeable {
+
+    private static final Logger logger = LoggerFactory.getLogger(ProactiveOpenConnectionsProcessor.class);
+    private Sinks.Many<OpenConnectionTask> openConnectionsTaskSink;
+    private Sinks.Many<OpenConnectionTask> openConnectionsTaskSinkBackUp;
+    private final ConcurrentHashMap<String, List<OpenConnectionTask>> endpointsUnderMonitorMap;
+    private final AtomicReference<AsyncDocumentClient.OpenConnectionAggressivenessHint> aggressivenessHint;
+    private static final Map<AsyncDocumentClient.OpenConnectionAggressivenessHint, ConcurrencyConfiguration> concurrencySettings = new HashMap<>();
+    private final IOpenConnectionsHandler openConnectionsHandler;
+    private final RntbdEndpoint.Provider endpointProvider;
+    private Disposable openConnectionBackgroundTask;
+
+    static {
+        concurrencySettings.put(AsyncDocumentClient.OpenConnectionAggressivenessHint.DEFENSIVE, new ConcurrencyConfiguration(Configs.getOpenConnectionsDefensiveConcurrency(), Configs.getOpenConnectionsDefensiveConcurrency()));
+        concurrencySettings.put(AsyncDocumentClient.OpenConnectionAggressivenessHint.AGGRESSIVE, new ConcurrencyConfiguration(Configs.getOpenConnectionsAggressiveConcurrency(), Configs.getOpenConnectionsAggressiveConcurrency()));
+    }
+
+    public ProactiveOpenConnectionsProcessor(final RntbdEndpoint.Provider endpointProvider, Duration aggressiveConnectionEstablishmentDuration) {
+        this.openConnectionsTaskSink = Sinks.many().multicast().onBackpressureBuffer();
+        this.openConnectionsTaskSinkBackUp = Sinks.many().multicast().onBackpressureBuffer();
+        this.aggressivenessHint = new AtomicReference<>(AsyncDocumentClient.OpenConnectionAggressivenessHint.AGGRESSIVE);
+        this.endpointsUnderMonitorMap = new ConcurrentHashMap<>();
+        this.openConnectionsHandler = new RntbdOpenConnectionsHandler(endpointProvider);
+        this.endpointProvider = endpointProvider;
+
+        // start as a background task
+        openConnectionBackgroundTask = this.getOpenConnectionsPublisher();
+        Flux<Integer> backgroundOpenConnectionsSinkReInstantiationTask = Flux
+                .just(1)
+                .delayElements(aggressiveConnectionEstablishmentDuration);
+
+        if (aggressiveConnectionEstablishmentDuration != null && aggressiveConnectionEstablishmentDuration.compareTo(Duration.ZERO) > 0) {
+            backgroundOpenConnectionsSinkReInstantiationTask
+                    .doOnComplete(() -> reInstantiateOpenConnectionsPublisherAndSubscribe())
+                    .subscribeOn(CosmosSchedulers.OPEN_CONNECTIONS_BOUNDED_ELASTIC)
+                    .subscribe();
+        }
+    }
+
+    public ProactiveOpenConnectionsProcessor(final RntbdEndpoint.Provider endpointProvider) {
+        this(endpointProvider, null);
+    }
+
+    public synchronized OpenConnectionTask submitOpenConnectionTaskOutsideLoop(
```

Comment:
every method is synchronized in this class, this is not good. We should have synchronized on granular level. 
Also, synchronized methods should be very quick, and should only change a state which is shared by other threads. 
This has high complexity, and can cause problems, specially with use of Flux / Mono - reactor in general.

Reply:
Resolved it offline.